app/controllers/sessions_controller.rb

```diff
@@ -25,12 +25,44 @@ def create
   end
 
   def twitter
+    add_twitter = session[:add_twitter]
+    session.delete :add_twitter
+
     auth = request.env["omniauth.auth"]
-    user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
+
+    # sign in via twitter
+    user = User.where(uid: auth["uid"]).first
     if user
-      session[:user_id] = user.id
-      flash[:notice] = "You have been logged in through Twitter."
-      redirect_back_or root_url
+      if add_twitter
+        # trying to add a twitter uid, but another user is already using it
+        flash[:alert] = "You've already created a shrtnr account using this Twitter id"
+        redirect_to settings_url #, error: "You've already created a shrtnr account using this Twitter id"
+      else
+        session[:user_id] = user.id
+        flash[:notice] = "You have been logged in through Twitter."
+        redirect_back_or root_url
+      end
+    else
+      if add_twitter
+        # adding twitter credentials to the current user
+        user = current_user
+        if user
```

Comment:
Since we're only checking if there is a current_user (that the user is `#signed_in?`), and the user *has* to be signed in to set `session[:add_twitter]` to true, you could probably drop the entire redirect. That way, if a user hits `/auth/twitter` (like the link to add the uid does), we can check if the user is `signed_in?` and then perform this block.